### DISCUSSION TITLE
behind-the-scenes-of-collection-expressions-part-1-introducing-collection-expressions-in-csharp12/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#423] [Announcements]
# behind-the-scenes-of-collection-expressions-part-1-introducing-collection-expressions-in-csharp12/

In this post I introduce C#12 collection expressions, discuss the benefits of the unified syntax for refactoring, and show how to use the spread element

https://andrewlock.net/behind-the-scenes-of-collection-expressions-part-1-introducing-collection-expressions-in-csharp12/

**Comment by lizanx** [comment]
Great post!

A minor typo at the beginning:
> For example, this creates a new List<int> and initializes it with 4 values: `var values = new List<string> { "1", "2", "3", "4", "5" };`

It should be "5 values".

**Comment by modestotech** [comment]
Great stuff! Looking forward to reading the next part of the series. Small stuff yes but will remove a bunch of boiler plate. 

Question on below, why should you never do the first way? Because it looks worse or is it less performant?
```
var empty = new int[]{}; // You should generally never do this...
var empty = Array.Empty<int>(); // ...instead, prefer this!
```

**Reply by andrewlock** [reply]
The first way allocates a new array every time, whereas the second one reuses an existing empty array. Given the array is empty (so can't be modified) creating a new one every time just puts unnecessary pressure on the GC 🙂

**Reply by modestotech** [reply]
Thanks for the explanation, Andrew, I have been using both options interchangeably without reflecting more on it, now I know what to use, until the `[]` becomes available :)

**Reply by Nitload-NSI** [reply]
Additional, now Roslyn will recommend the fallowing style:

``` csharp

int[] empty = [];    // recommended
int[] empty = Array.Empty<int>();     //Old style, IDE301 info

```
The two expressions above are same in IL
